test/integration/xds_config_tracker_integration_test.cc

```diff
@@ -0,0 +1,291 @@
+#include <atomic>
+
+#include "envoy/config/route/v3/route.pb.h"
+#include "envoy/config/subscription.h"
+#include "envoy/config/xds_config_tracker.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+#include "envoy/service/runtime/v3/rtds.pb.h"
+
+#include "test/common/grpc/grpc_client_integration.h"
+#include "test/config/v2_link_hacks.h"
+#include "test/integration/http_integration.h"
+#include "test/integration/xds_config_tracker_test.pb.h"
+#include "test/test_common/registry.h"
+#include "test/test_common/utility.h"
+
+#include "absl/synchronization/mutex.h"
+#include "gtest/gtest.h"
+
+using testing::HasSubstr;
+
+namespace Envoy {
+namespace {
+
+constexpr char XDS_CLUSTER_NAME[] = "xds_cluster";
+
+class TestXdsConfigTracker : public Config::XdsConfigTracker {
+public:
+  TestXdsConfigTracker() {
+    ReceiveCount = 0;
+    IngestedCount = 0;
+    FailureCount = 0;
+    ErrorMessage = "";
+  }
+
+  void onConfigIngested(const absl::string_view,
+                        const std::vector<Config::DecodedResourcePtr>&) override {
+    ++IngestedCount;
+  }
+
+  void onConfigIngested(const absl::string_view, const std::vector<Config::DecodedResourcePtr>&,
+                        const Protobuf::RepeatedPtrField<std::string>&) override {
+    ++IngestedCount;
+  }
+
+  void onConfigReceivedOrFailed(const envoy::service::discovery::v3::DiscoveryResponse&,
+                                const Config::ProcessingDetails& process_details) override {
+    countState(process_details.state_);
+    if (process_details.state_ == Config::ProcessingState::FAILED) {
+      ErrorMessage = process_details.error_detail_.message();
+    }
+  }
+
+  void onConfigReceivedOrFailed(const envoy::service::discovery::v3::DeltaDiscoveryResponse&,
+                                const Config::ProcessingDetails& process_details) override {
+    countState(process_details.state_);
+    if (process_details.state_ == Config::ProcessingState::FAILED) {
+      ErrorMessage = process_details.error_detail_.message();
+    }
+  }
+
+  static std::atomic<int> ReceiveCount;
+  static std::atomic<int> IngestedCount;
+  static std::atomic<int> FailureCount;
+  static std::string ErrorMessage;
+
+private:
+  void countState(const Config::ProcessingState& state) {
+    switch (state) {
+    case Config::ProcessingState::RECEIVED:
+      ++ReceiveCount;
+      break;
+    case Config::ProcessingState::FAILED:
+      ++FailureCount;
+      break;
+    };
+  }
+};
+
+std::atomic<int> TestXdsConfigTracker::ReceiveCount;
```

Comment:
I think this can be done by introducing these as test counters (see https://github.com/envoyproxy/envoy/blob/f4144e8b02181d927fe4bea5eba0e95316b5ef7a/test/integration/filters/test_network_filter.cc#L14 as an example).

Reply:
Thanks! Added the stats counter for the test.